src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -421,9 +425,45 @@ entails SolverOptions{..} constraint context hints =
               let fields = [ ("reflectType", Abs (VarBinder nullSourceSpan UnusedIdent) (asExpression ref)) ] in
               pure $ App (Constructor nullSourceSpan (coerceProperName . dictTypeName <$> C.Reflectable)) (Literal nullSourceSpan (ObjectLiteral fields))
 
-            unknownsInAllCoveringSets :: [SourceType] -> S.Set (S.Set Int) -> Bool
-            unknownsInAllCoveringSets tyArgs = all (\s -> any (`S.member` s) unkIndices)
-              where unkIndices = findIndices containsUnknowns tyArgs
+            unknownsInAllCoveringSets :: [(Ident, SourceType, Maybe [[Text]])] -> [SourceType] -> S.Set (S.Set Int) -> UnknownsHint
+            unknownsInAllCoveringSets tyClassMembers tyArgs converingSets = do
+              let unkIndices = findIndices containsUnknowns tyArgs
+              if all (\s -> any (`S.member` s) unkIndices) converingSets then 
+                fromMaybe Unknowns unknownsRequiringVtas
+              else 
+                NoUnknowns
+              where
+                unknownsRequiringVtas = do
+                  tyClassModuleName <- getQual className'
+                  let
+                    tyClassMemberVta :: M.Map (Qualified Ident) [[Text]]
+                    tyClassMemberVta = M.fromList $ mapMaybe qualifyAndFilter tyClassMembers
+                      where
+                        -- Only keep type class members that need VTAs to resolve their type class instances
+                        qualifyAndFilter (ident, _, mbVtas) = (Qualified (ByModuleName tyClassModuleName) ident,) <$> mbVtas
+
+                    tyClassMembersInExpr :: Expr -> [(Qualified Ident, [[Text]])]
+                    tyClassMembersInExpr = getVars
+                      where
+                        (_, getVars, _, _, _) = everythingOnValues (++) ignore getVarIdents ignore ignore ignore
+                        ignore = const []
+                        getVarIdents = \case
+                          Var _ ident | Just vtas <- M.lookup ident tyClassMemberVta -> 
+                            [(ident, vtas)]
+                          _ -> 
+                            []
+
+                    findTyClassMembersInExprHint
+                      :: ErrorMessageHint
+                      -> Maybe [(Qualified Ident, [[Text]])]
+                      -> Maybe [(Qualified Ident, [[Text]])]
+                    findTyClassMembersInExprHint next acc = case (acc, next) of
+                      (Nothing, ErrorCheckingType expr _) -> Just $ tyClassMembersInExpr expr
+                      (_, _) -> acc
+                  
+                  tyClassMembers' <- foldr findTyClassMembersInExprHint Nothing hints
```

Comment:
```suggestion
                    getECTExpr = \case
                      ErrorCheckingType expr _ -> Just expr
                      _ -> Nothing
                      
                  tyClassMembers' <- headMay $ mapMaybe (fmap tyClassMembersInExpr . getECTExpr) hints
```

Reply:
Are there ever multiple `ErrorCheckingType` values in `hints`? When we store the `hints` data constructor, we call `getHints` which takes the environment's hints and reverses them. I used `foldr` here to ensure I got the most recent `ErrorCheckingType` in that stack (i.e. the value that would be closer to the head of the list before the list is reversed rather than the tail of the the list before the list is reversed).

If there are, then I think this could should still use `foldr`. If not, then your code is cleaner.

Reply:
I think I'll just commit this code and see whether the tests pass.